### PR TITLE
Provide markdown details for CheckRun.

### PR DESCRIPTION
**Pull request by tmcgilchrist**
Fill in more of the CheckRun payload to support actions, output markdown and Skipped conclusion type.

### CODE REVIEW

File: plugins/github/api.ml [hunk 1]
```diff
@@ -94,35 +94,71 @@ module Status = struct
...
+  let action_to_json {label; description; identifier} =
+    `Assoc [ ("label", `String label)
+           ; ("description", `String description)
+           ; ("identifier", `String identifier)]
+
+  let v ?text ?summary ?url ?(actions=[]) state =
+    (* A maximum of three actions are accepted by GitHub. *)
+    { state; summary; text; actions; url }
+
+  let state_to_string description summary actions = function
```
Comment (talex5) [inline plugins/github/api.ml:119]: This is badly named; it doesn't return a string!

**Review by talex5 — COMMENTED** [review]
Looks OK to me.

The `state_to_string` function is a bit hard to check though. Perhaps it would be clearer written something like this:

```ocaml
  let state_json ~title ~summary ?description ~status () =
    let output = [ ("title", `String title)
                 ; ("summary", `String summary) ] @
                 (match description with None -> [] | Some x -> ["text", `String x])
    in
    [ "status", `String status
    ; "output", `Assoc output]

  let state_to_string description summary actions = function
    | `Queued -> 
      state_json ()
        ~title:"Queued"
        ~summary:(Option.value summary ~default:"Queued")
        ~status:"queued"
        ~description
    ...
```

File: plugins/github/api.ml (hunk 1 above)
Comment (tmcgilchrist) [reply plugins/github/api.ml]: Reworked this.

File: plugins/github/api.ml [hunk 2]
```diff
@@ -94,36 +94,68 @@ module Status = struct
...
 
-  let json_items { state; description; url } =
-    state_to_string state @
-    (match description with None -> [] | Some x -> ["description", `String x]) @
-    (match url with None -> [] | Some x -> ["details_url", `String (Uri.to_string x)])
+    | `Completed `Success -> 
+      ("conclusion", `String "success") ::
+      state_json ~title:"Completed" 
+                 ~summary:(Option.value summary ~default:"Completed")
+                 ~status:"completed" ~text ()
```
Comment (talex5) [inline plugins/github/api.ml:144]: This is OK, but note that doing it this way prevents using automatic indentation (e.g. with ocp-indent). The original style I suggested avoids that problem:
```ocaml
      state_json ()
        ~title:"Queued"
        ~summary:(Option.value summary ~default:"Queued")
        ~status:"queued"
        ~text
```

File: plugins/github/api.ml [hunk 3]
```diff
@@ -94,36 +94,68 @@ module Status = struct
...
+
+    | `Completed (`Failure _) ->
+       [ "conclusion", `String "failure"
```
Comment (talex5) [inline plugins/github/api.ml:147]: This could do with being run through an auto-indenter ;-)

File: plugins/github/api.ml [hunk 4]
```diff
@@ -94,36 +94,68 @@ module Status = struct
...
+       ; "actions", `List (List.map action_to_json actions) ] @ 
+       state_json ~title:"Failure" 
+                  ~summary:(Option.value summary ~default:"Failure")
+                  ~status:"completed" ~text ()
+
+    | `Completed (`Skipped _) -> 
+      [ "conclusion", `String "skipped"
```
Comment (talex5) [inline plugins/github/api.ml:154]: BTW, the idea with the `state_json` was to keep all the magic strings in one place. I just didn't read far enough to see that there were more. Perhaps it should have been:

```ocaml
  let state_json ~title ~summary ~text ~status ?conclusion ?(actions=[]) () =
    let output = [ ("title", `String title)
                 ; ("summary", `String summary) ] @
                 (match text with None -> [] | Some x -> ["text", `String x])
    in
    [ "status", `String status
    ; "output", `Assoc output ] @
    (if actions = [] then [] else ["actions", `List (List.map action_to_json actions)]) @
    (match conclusion with None -> [] | Some x -> ["conclusion", `String x])
```

**Review by talex5 — APPROVED** [review]
This looks fine. I made some more style comments, but these are just suggestions. Feel free to merge with or without them.

**Pull request merged**